miner/worker.go

```diff
@@ -360,619 +238,200 @@ func (w *worker) createTxCmp() func(tx1 *types.Transaction, tx2 *types.Transacti
-		}
-	}
-}
-
-// taskLoop is a standalone goroutine to fetch sealing task from the generator and
-// push them to consensus engine.
-func (w *worker) taskLoop() {
-
-	for {
-		select {
-		case task := <-w.taskCh:
-			if w.newTaskHook != nil {
-				w.newTaskHook(task)
-			}
-
-			if w.skipSealHook != nil && w.skipSealHook(task) {
-				continue
-			}
-
-			if err := w.engine.Seal(w.chain, task.block); err != nil {
-				log.Warn("Block sealing failed", "err", err)
-			}
-		case <-w.exitCh:
-			return
 		}
 	}
-}
-
-// makeCurrent creates a new environment for the current cycle.
-func (w *worker) makeCurrent(parent *types.Block, header *types.Header) error {
-	state, err := w.chain.StateAt(parent.Root())
-	if err != nil {
-		return err
-	}
 
-	env := &environment{
-		signer:    types.NewEIP155Signer(w.chainConfig.ChainID),
-		state:     state,
-		ancestors: mapset.NewSet(),
-		header:    header,
-		gasLimit:  core.CalcGasLimit(parent, state),
-	}
-
-	// when 08 is processed ancestors contain 07 (quick block)
-	for _, ancestor := range w.chain.GetBlocksFromHash(parent.Hash(), 7) {
-		env.ancestors.Add(ancestor.Hash())
-	}
-
-	// Keep track of transactions which return errors so they can be removed
-	env.tcount = 0
-	w.current = env
-	return nil
 }
 
-// updateSnapshot updates pending snapshot block and state.
-// Note this function assumes the current variable is thread safe.
-func (w *worker) updateSnapshot() {
-	w.snapshotMu.Lock()
-	defer w.snapshotMu.Unlock()
-
-	w.snapshotBlock = types.NewBlock(
-		w.current.header,
-		w.current.txs,
-		w.current.receipts,
-		w.current.randomness,
-	)
-
-	w.snapshotState = w.current.state.Copy()
-}
-
-func (w *worker) commitTransaction(tx *types.Transaction, txFeeRecipient common.Address) ([]*types.Log, error) {
-	snap := w.current.state.Snapshot()
-
-	receipt, err := core.ApplyTransaction(w.chainConfig, w.chain, &txFeeRecipient, w.current.gasPool, w.current.state, w.current.header, tx, &w.current.header.GasUsed, *w.chain.GetVMConfig())
-	if err != nil {
-		w.current.state.RevertToSnapshot(snap)
-		return nil, err
-	}
-	w.current.txs = append(w.current.txs, tx)
-	w.current.receipts = append(w.current.receipts, receipt)
-
-	return receipt.Logs, nil
-}
-
-func (w *worker) commitTransactions(txs *types.TransactionsByPriceAndNonce, txFeeRecipient common.Address, interrupt *int32) bool {
-	// Short circuit if current is nil
-	if w.current == nil {
-		return true
-	}
-
-	if w.current.gasPool == nil {
-		w.current.gasPool = new(core.GasPool).AddGas(w.current.gasLimit)
-	}
+// mainLoop is a standalone goroutine to create tasks and submit to the engine.
+func (w *worker) mainLoop() {
+	defer w.chainHeadSub.Unsubscribe()
+	defer w.txsSub.Unsubscribe()
+	// Context and cancel function for the currently executing block construction
+	// Cancel needs to be called in each exit path to make the linter happy
+	// because go struggles with analyzing lexical scoping.
+	var taskCtx context.Context
+	var cancel context.CancelFunc
```

Comment:
suggest you add:
```go
defer func() {
			if cancel != nil {
				cancel()
			}  
}()
```

here, and remove the checks on "stop". seems simpler

Reply:
It would be, but golang cannot recognize that form as equivalent and the linter complains.

Reply:
Minimal example showing go can't handle this: https://play.golang.org/p/dwrvFNabLhm